clustercontroller-core/src/test/java/com/yahoo/vespa/clustercontroller/core/EventDiffCalculatorTest.java

```diff
@@ -14,80 +16,94 @@
         }
 
         EventFixture clusterStateBefore(String stateStr) {
-            clusterStateBefore = ClusterState.stateFromString(stateStr);
+            baselineBefore.clusterState(stateStr);
             return this;
         }
         EventFixture clusterStateAfter(String stateStr) {
-            clusterStateAfter = ClusterState.stateFromString(stateStr);
+            baselineAfter.clusterState(stateStr);
             return this;
         }
-        EventFixture storageNodeReasonBefore(int index, NodeStateReason reason) {
-            nodeReasonsBefore.put(storageNode(index), reason);
+        EventFixture storageNodeReasonBefore(int nodeIndex, NodeStateReason reason) {
+            baselineBefore.storageNodeReason(nodeIndex, reason);
             return this;
         }
-        EventFixture storageNodeReasonAfter(int index, NodeStateReason reason) {
-            nodeReasonsAfter.put(storageNode(index), reason);
+        EventFixture storageNodeReasonAfter(int nodeIndex, NodeStateReason reason) {
+            baselineAfter.storageNodeReason(nodeIndex, reason);
             return this;
         }
         EventFixture clusterReasonBefore(ClusterStateReason reason) {
-            this.clusterReasonBefore = Optional.of(reason);
+            baselineBefore.clusterReason(reason);
             return this;
         }
         EventFixture clusterReasonAfter(ClusterStateReason reason) {
-            this.clusterReasonAfter = Optional.of(reason);
+            baselineAfter.clusterReason(reason);
             return this;
         }
         EventFixture currentTimeMs(long timeMs) {
             this.currentTimeMs = timeMs;
             return this;
         }
+        EventFixture derivedClusterStateBefore(String bucketSpace, String stateStr) {
+            getBuilder(derivedBefore, bucketSpace).clusterState(stateStr);
+            return this;
+        }
+        EventFixture derivedClusterStateAfter(String bucketSpace, String stateStr) {
+            getBuilder(derivedAfter, bucketSpace).clusterState(stateStr);
+            return this;
+        }
+        EventFixture derivedStorageNodeReasonBefore(String bucketSpace, int nodeIndex, NodeStateReason reason) {
+            getBuilder(derivedBefore, bucketSpace).storageNodeReason(nodeIndex, reason);
+            return this;
+        }
+        EventFixture derivedStorageNodeReasonAfter(String bucketSpace, int nodeIndex, NodeStateReason reason) {
+            getBuilder(derivedAfter, bucketSpace).storageNodeReason(nodeIndex, reason);
+            return this;
+        }
+        private static AnnotatedClusterState.Builder getBuilder(Map<String, AnnotatedClusterState.Builder> derivedStates, String bucketSpace) {
+            AnnotatedClusterState.Builder result = derivedStates.get(bucketSpace);
+            if (result == null) {
+                result = new AnnotatedClusterState.Builder();
+                derivedStates.put(bucketSpace, result);
+            }
```

Comment:
Can consider using `putIfAbsent` instead of explicit `get`+`put` pair

Reply:
The problem with putIfAbsent is that it returns null if the specified key is not already associated with a value (and then associated with the given value). I haven't found a one-liner that is feasible here.

Reply:
computeIfAbsent

Reply:
Thanks @bratseth